src/librbd/crypto/FormatRequest.cc

```diff
@@ -40,9 +41,25 @@ void FormatRequest<I>::send() {
     return;
   }
 
-  if (m_image_ctx->crypto == nullptr) {
+  m_image_ctx->image_lock.lock_shared();
+  bool is_encryption_loaded = m_image_ctx->encryption_format.get() != nullptr;
+  bool is_formatted_clone = m_image_ctx->is_formatted_clone;
+  m_image_ctx->image_lock.unlock_shared();
+
+  if (is_formatted_clone) {
+    lderr(m_image_ctx->cct) << "cloned image already formatted" << dendl;
+    finish(-EINVAL);
+    return;
+  }
+
+  if (!is_encryption_loaded) {
     format();
     return;
+  } else if (m_image_ctx->parent != nullptr) {
+    lderr(m_image_ctx->cct) << "cannot thin format while encryption is loaded"
+                            << dendl;
+    finish(-EINVAL);
```

Comment:
Well as I noted in a previous comment, thin-formatting semantics is different then flat formatting.
If the image was already formatted, it means IO already happened on the image, i.e. copyups, and you're no longer basing on a "clean parent".
The user might mistakenly think that, so we dis-allow it.
If you want to re-format a clone, you have to re-clone.

Reply:
Following this logic, the user might mistakenly think re-formatting a flat image would get them a clean (empty) flat image.

I realize that I'm playing a bit of the devil's advocate but I'm really trying to make the semantics crystal clear.  I think this check is yet another thing that ties with the "is I/O between cloning and formatting allowed" question.  If it is, failing with `EINVAL` here is meaningless.

Reply:
I think it is better this way to avoid user confusion due to the fact that encryption is not loaded following the format of a cloned image.
In other words, the crypto won't be fully loaded after format completes, and this check makes it more clear to the user, since encryption is guaranteed to be not loaded in the first place.

Reply:
I see your point.  Let's keep it but please reword the message -- "thin format" is no longer a thing, use "cloned image" instead.